// @elastic/synthetics/dist/index.js
"use strict";
/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.request = exports.devices = exports.mfa = exports.expect = exports.after = exports.before = exports.afterAll = exports.beforeAll = exports.monitor = exports.step = exports.journey = exports.run = void 0;
const globals_1 = require("./core/globals");
async function run(options) {
    return globals_1.runner._run(options);
}
exports.run = run;
/**
 * Export all core module functions
 */
var core_1 = require("./core");
Object.defineProperty(exports, "journey", { enumerable: true, get: function () { return core_1.journey; } });
Object.defineProperty(exports, "step", { enumerable: true, get: function () { return core_1.step; } });
Object.defineProperty(exports, "monitor", { enumerable: true, get: function () { return core_1.monitor; } });
Object.defineProperty(exports, "beforeAll", { enumerable: true, get: function () { return core_1.beforeAll; } });
Object.defineProperty(exports, "afterAll", { enumerable: true, get: function () { return core_1.afterAll; } });
Object.defineProperty(exports, "before", { enumerable: true, get: function () { return core_1.before; } });
Object.defineProperty(exports, "after", { enumerable: true, get: function () { return core_1.after; } });
var expect_1 = require("./core/expect");
Object.defineProperty(exports, "expect", { enumerable: true, get: function () { return expect_1.expect; } });
exports.mfa = __importStar(require("./core/mfa"));
var playwright_core_1 = require("playwright-core");
Object.defineProperty(exports, "devices", { enumerable: true, get: function () { return playwright_core_1.devices; } });
Object.defineProperty(exports, "request", { enumerable: true, get: function () { return playwright_core_1.request; } });
