// @elastic/synthetics/dist/dsl/step.js
"use strict";
/**
 * MIT License
 *
 * Copyright (c) 2020-present, Elastic NV
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 *
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.Step = void 0;
class Step {
    name;
    index;
    location;
    #cb;
    skip = false;
    soft = false;
    only = false;
    _startTime = 0;
    duration = -1;
    status = 'pending';
    error;
    url;
    constructor(name, index, cb, location) {
        this.name = name;
        this.index = index;
        this.#cb = cb;
        this.location = location;
    }
    get cb() {
        return this.#cb;
    }
    /**
     * @deprecated Since version 1.17.0. Please do not rely on the internal methods.
     * Alias cb for backwards compatibility
     */
    get callback() {
        return this.#cb;
    }
}
exports.Step = Step;
